src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -799,12 +798,16 @@ public Solution AddAnalyzerReferences(ProjectId projectId, IEnumerable<AnalyzerR
         foreach (var analyzerReference in collection)
         {
             if (this.SolutionState.ContainsAnalyzerReference(projectId, analyzerReference))
-            {
                 throw new InvalidOperationException(WorkspacesResources.The_project_already_contains_the_specified_reference);
-            }
         }
 
-        return WithCompilationState(CompilationState.AddAnalyzerReferences(this.SolutionState.AddAnalyzerReferences(projectId, collection), collection));
```

Comment:
we only expose Add/Remove at the top level.  internally, these all boil down to 'With' calls.  this makes it so we have a single chokepoint to add some new logic.